### PR TITLE
Upgrade horizontal scaling information (Keda)

### PR DESCRIPTION
**Pull request by Richard87**
Example response:

Note: Limitations in Keda status/hpa status fields causes  resource types to match on the first type (so the code below will be invalid for duplicate cpu and memory triggers (only the first of each type is correct, the other is duplicated)

```json
        "horizontalScalingSummary": {
          "minReplicas": 1,
          "maxReplicas": 3,
          "cooldownPeriod": 300,
          "pollingInterval": 30,
          "triggers": [
            {
              "name": "cpu4",
              "current_utilization": "1",
              "target_utilization": "50",
              "type": "cpu"
            },
            {
              "name": "cpu4",
              "current_utilization": "1",
              "target_utilization": "50",
              "type": "cpu"
            },
            {
              "name": "memory2",
              "current_utilization": "11",
              "target_utilization": "50",
              "type": "memory"
            },
            {
              "name": "cron",
              "current_utilization": "1",
              "target_utilization": "1",
              "type": "cron"
            },
            {
              "name": "azuresb",
              "current_utilization": "",
              "target_utilization": "10",
              "type": "azure-servicebus",
              "error": "Failing: number of failurs: 7020"
            }
          ],
          "currentCPUUtilizationPercentage": 1,
          "targetCPUUtilizationPercentage": 50,
          "currentMemoryUtilizationPercentage": 11,
          "targetMemoryUtilizationPercentage": 50
        },
```

### CODE REVIEW

File: api/utils/kubernetes.go [hunk 1]
```diff
@@ -108,6 +110,9 @@ func getOutClusterClientConfig(token string, impersonation radixmodels.Impersona
 
 		kubeConfig.Impersonate = impersonationConfig
 	}
+	kubeConfig.Wrap(logs.Logger(func(e *zerolog.Event) {
```
Comment (nilsgstrabo) [inline api/utils/kubernetes.go:113]: Should we make roundtripper logging configurable via options?

File: api/utils/kubernetes.go [hunk 2]
```diff
@@ -121,6 +126,9 @@ func getInClusterClientConfig(options []RestClientConfigOption) *restclient.Conf
 			log.Fatal().Err(err).Msg("getClusterConfig InClusterConfig")
 		}
 	}
+	config.Wrap(logs.Logger(func(e *zerolog.Event) {
```
Comment (nilsgstrabo) [inline api/utils/kubernetes.go:129]: Should we make roundtripper logging configurable via options?

**Review by nilsgstrabo — CHANGES_REQUESTED** [review]
Added some suggestions

File: api/deployments/component_handler.go [hunk 3]
```diff
@@ -155,12 +99,16 @@ func GetComponentStateFromSpec(
 	deployment *deploymentModels.DeploymentSummary,
 	deploymentStatus v1.RadixDeployStatus,
 	environmentConfig v1.RadixCommonEnvironmentConfig,
-	component v1.RadixCommonDeployComponent) (*deploymentModels.Component, error) {
+	component v1.RadixCommonDeployComponent,
```
[the remark this answers is not in the document]
Comment (Richard87) [reply api/deployments/component_handler.go]: fixed

File: api/environments/environment_handler.go [hunk 4]
```diff
@@ -449,12 +452,21 @@ func (eh EnvironmentHandler) getRadixCommonComponentUpdater(ctx context.Context,
 		updater = &radixDeployJobComponentUpdater{base: baseUpdater}
 	}
 
+	hpas, err := eh.getHPAsInEnvironment(ctx, appName, envName)
```
[the remark this answers is not in the document]
Comment (Richard87) [reply api/environments/environment_handler.go:455]: fiexd with friends :)

File: api/environments/environment_handler.go [hunk 5]
```diff
@@ -449,12 +452,21 @@ func (eh EnvironmentHandler) getRadixCommonComponentUpdater(ctx context.Context,
...
+	if err != nil {
+		return nil, err
+	}
+	scalers, err := eh.getScaledObjectsInEnvironment(ctx, appName, envName)
```
[the remark this answers is not in the document]
Comment (Richard87) [reply api/environments/environment_handler.go:459]: fiexd with friends :)

File: api/environments/utils.go [hunk 6]
```diff
@@ -30,3 +32,23 @@ func (eh EnvironmentHandler) getRadixApplicationInAppNamespace(ctx context.Conte
...
+
+func (eh EnvironmentHandler) getScaledObjectsInEnvironment(ctx context.Context, appName, envName string) ([]v1alpha1.ScaledObject, error) {
+	envNs := operatorutils.GetEnvironmentNamespace(appName, envName)
+	scaledObjects, err := eh.accounts.UserAccount.KedaClient.KedaV1alpha1().ScaledObjects(envNs).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	return scaledObjects.Items, nil
+}
```
[the remark this answers is not in the document]
Comment (Richard87) [reply api/environments/utils.go:54]: Done

File: api/utils/kubernetes.go (hunk 2 above)
Comment (Richard87) [reply api/utils/kubernetes.go:129]: Can be activated by setting log level to trace :)

File: api/deployments/component_handler.go [hunk 7]
```diff
@@ -2,435 +2,78 @@ package deployments
...
-	selector := labelselector.ForAuxiliaryResource(appName, componentName, auxType).String()
-	deployments, err := kubeClient.AppsV1().Deployments(envNamespace).List(ctx, metav1.ListOptions{LabelSelector: selector})
+	eventList, err := kubequery.GetEventsForEnvironment(ctx, deploy.accounts.UserAccount.Client, appName, envName)
 	if err != nil {
 		return nil, err
 	}
-	if len(deployments.Items) == 0 {
-		auxResourceDeployment.Status = deploymentModels.ComponentReconciling.String()
-		return &auxResourceDeployment, nil
+	certs, err := kubequery.GetCertificatesForEnvironment(ctx, deploy.accounts.UserAccount.CertManagerClient, appName, envName)
```
Comment (nilsgstrabo) [inline api/deployments/component_handler.go:53]:
```suggestion
	certs, err := kubequery.GetCertificatesForEnvironment(ctx, deploy.accounts.ServiceAccount.CertManagerClient, appName, envName)
```
UserAccount does not have access (yet), so we must use ServiceAccount until fixed in radix-operator

File: api/deployments/component_handler.go [hunk 8]
```diff
@@ -2,435 +2,78 @@ package deployments
...
+	if err != nil {
+		return nil, err
 	}
-	deployment := deployments.Items[0]
-
-	pods, err := kubeClient.CoreV1().Pods(envNamespace).List(ctx, metav1.ListOptions{LabelSelector: selector})
+	certRequests, err := kubequery.GetCertificateRequestsForEnvironment(ctx, deploy.accounts.UserAccount.CertManagerClient, appName, envName)
```
Comment (nilsgstrabo) [inline api/deployments/component_handler.go:57]:
```suggestion
	certRequests, err := kubequery.GetCertificateRequestsForEnvironment(ctx, deploy.accounts.ServiceAccount.CertManagerClient, appName, envName)
```

**Review by nilsgstrabo — APPROVED** [review]

**Pull request merged**